File: python/cquant/api_server/schemas/knowledge.py
"""Knowledge-base API schemas."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


class IngestRequestBody(BaseModel):
    uri: str
    logical_type: Literal["research", "strategy", "notes", "data"] = "research"
    source_name: str = ""
    title: str = ""
    language: str = "zh-CN"


class IngestResponseBody(BaseModel):
    doc_id: str
    status: str              # 'ok' | 'duplicate' | 'error'
    chunk_count: int = 0
    run_id: str = ""
    error: str = ""


class SearchRequestBody(BaseModel):
    text: str
    top_k: int = 10
    mode: Literal["hybrid", "semantic", "keyword"] = "hybrid"
    logical_type: str | None = None
    source_name: str | None = None


class SearchHitBody(BaseModel):
    doc_id: str
    title: str = ""
    source_name: str = ""
    logical_type: str = ""
    score: float = 0.0
    headline: str = ""


class SearchResponseBody(BaseModel):
    hits: list[SearchHitBody]
    total_found: int
    latency_ms: int


class QARequestBody(BaseModel):
    question: str = Field(..., min_length=1, description="The question to answer using the knowledge base")
    top_k: int = Field(default=5, ge=1, le=20, description="Number of context snippets to retrieve")
    model: Literal["claude", "openai"] = Field(default="claude", description="LLM provider to use")


class QASourceBody(BaseModel):
    doc_id: str
    snippet: str = ""
    score: float = 0.0


class QAResponseBody(BaseModel):
    answer: str
    sources: list[QASourceBody]
    model: str

File: web/src/lib/api/knowledge.ts
/**
 * cQuant API — Knowledge base domain.
 */

import { api, type RequestConfig } from './client'

// ── Types (not yet in types/) ────────────────────────────────────────────────

export interface KnowledgeDoc {
  doc_id: string
  title: string
  source_name: string
  logical_type: string
  language: string
  ingested_at: string
}

export interface SearchHit {
  doc_id: string
  title: string
  source_name: string
  logical_type: string
  score: number
  headline: string
}

export interface SearchResponse {
  hits: SearchHit[]
  total_found: number
  latency_ms: number
}

// ── API ────────────────────────────────────────────────────────────────────

export const knowledgeApi = {
  list: (logicalType?: string, limit = 100, config?: RequestConfig) => {
    const params = new URLSearchParams({ limit: String(limit) })
    if (logicalType) params.set('logical_type', logicalType)
    return api.get<{ items: KnowledgeDoc[]; total: number }>(
      `/knowledge/docs?${params}`,
      config,
    )
  },

  get: (id: string, config?: RequestConfig) =>
    api.get<KnowledgeDoc>(`/knowledge/docs/${id}`, config),

  search: (text: string, topK = 10, config?: RequestConfig) =>
    api.post<SearchResponse>('/knowledge/search', { text, top_k: topK }, config),

  ingest: (
    body: { uri: string; logical_type?: string; source_name?: string; title?: string },
    config?: RequestConfig,
  ) => api.post('/knowledge/ingest', body, config),
}
